### PR TITLE
[SMTChecker] Keep knowledge about string literals

### LINKED ISSUE #6052: [SMTChecker] Support conversion from string to bytes

### PR DESCRIPTION
**Pull request by axic**
Fixes #6052.

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -834,7 +834,21 @@ void SMTEncoder::endVisit(Literal const& _literal)
...
 	else if (smt::isStringLiteral(type))
+	{
 		createExpr(_literal);
+
+		// Add constraints for the length and values as it is known.
+		auto symbArray = dynamic_pointer_cast<smt::SymbolicArrayVariable>(m_context.expression(_literal));
+		solAssert(symbArray, "");
+
+		auto value = _literal.value();
+		m_context.addAssertion(symbArray->length() == smtutil::Expression(value.length()));
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		m_context.addAssertion(symbArray->length() == value.length());
```
Does that work?

Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: Yes, can simplify both assertions -- well need to check the tests if it does not mess it up :)

**Review by leonardoalt — DISMISSED** [review]
Really nice!
Could add some tests that check the value of the element also, since that's also implemented

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: Seems to work, pushed.

**Comment by axic** [comment]
> Could add some tests that check the value of the element also, since that's also implemented

There is a test under `operators` for `bytes`. `string` can not be indexed, that's why I added the roundtrip conversion.

I could add a test similar to the bytes indexing with that roundtrip conversion.

**Review by leonardoalt — DISMISSED** [review]

**Comment by leonardoalt** [comment]
> I could add a test similar to the bytes indexing with that roundtrip conversion.

Yea that'd be good

**Comment by axic** [comment]
> Yea that'd be good

It is included, but it seems to trigger the z3 bug again :cry:

**Comment by leonardoalt** [comment]
Yea.. I'm compiling locally to see if I can reproduce it.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**